### ISSUE TITLE
Terraform 13/14: fails to download newer provider version if older version exists locally

### ISSUE THREAD

**Issue by bhadrim** [hashicorp/terraform#28818]
### Terraform Version

```
0.14.11
0.13.7
```

### Terraform Configuration Files

```
terraform {
  required_providers {
    google = {
      source = "hashicorp/google"
      version = "3.69.0"
    }
  }
  required_version = ">= 0.13"
}
```

### Debug Output

```
Tue May 25 2021 20:08:35 GMT+0000 (Coordinated Universal Time) 
Running terraform init ... 2021/05/25 20:08:35 
[INFO] Terraform version: 0.14.11 2021/05/25 20:08:35 
[INFO] Go runtime version: go1.15.6 2021/05/25 20:08:35 
[INFO] CLI args: []string{"/home/terraform/bin/terraform_0.14.11", "init", "-backend=false", "-input=false", "-get=true"} 2021/05/25 20:08:35 
[DEBUG] Attempting to open CLI config file: /stacks/stackjobs/temp_a29c2780-bd78-11eb-b4e9-b591e3e27225/.terraformrc 2021/05/25 20:08:35 Loading CLI configuration from /stacks/stackjobs/temp_a29c2780-bd78-11eb-b4e9-b591e3e27225/.terraformrc 2021/05/25 20:08:35 
[DEBUG] checking for credentials in "/stacks/stackjobs/temp_a29c2780-bd78-11eb-b4e9-b591e3e27225/.terraform.d/plugins" 2021/05/25 20:08:35 
[DEBUG] ignoring non-existing provider search directory terraform.d/plugins 2021/05/25 20:08:35 
[DEBUG] will search for provider plugins in /stacks/stackjobs/temp_a29c2780-bd78-11eb-b4e9-b591e3e27225/.terraform.d/plugins 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: /stacks/stackjobs/temp_a29c2780-bd78-11eb-b4e9-b591e3e27225/.terraform.d/plugins is a symlink to /home/terraform/.terraform.d/plugins 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.ibm.com/cam/camc v0.2.5 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.ibm.com/cam/camc/0.2.5/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.ibm.com/cam/ibmworkloaddeployer v1.0.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.ibm.com/cam/ibmworkloaddeployer/1.0.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.ibm.com/cam/icp-elk v0.2.3 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.ibm.com/cam/icp-elk/0.2.3/linux_amd64 2021/05/25 20:08:35 [TRACE] getproviders.SearchLocalDirectory: found registry.ibm.com/cam/nsxv v1.0.1 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.ibm.com/cam/nsxv/1.0.1/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.ibm.com/cam/ucd v0.5.1 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.ibm.com/cam/ucd/0.5.1/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/aliyun/alicloud v1.121.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/aliyun/alicloud/1.121.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/aws v2.33.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/aws/2.33.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/aws v2.44.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/aws/2.44.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/aws v3.36.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/aws/3.36.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/azurerm v1.35.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/azurerm/1.35.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/azurerm v2.55.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/azurerm/2.55.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/external v1.2.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/external/1.2.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/external v2.1.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/external/2.1.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/google v2.17.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/google/2.17.0/linux_amd64 2021/05/25 20:08:35 
[WARN] failed to read metadata about /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/google/3.64.0/linux_amd64: stat /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/google/3.64.0/linux_amd64: too many levels of symbolic links 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/http v1.1.1 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/http/1.1.1/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/http v2.1.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/http/2.1.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/kubernetes v1.9.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/kubernetes/1.9.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/kubernetes v2.0.3 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/kubernetes/2.0.3/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/local v1.4.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/local/1.4.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/local v2.1.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/local/2.1.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/null v2.1.2 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/null/2.1.2/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/null v3.1.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/null/3.1.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/random v2.2.1 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/random/2.2.1/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/random v3.1.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/random/3.1.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/tls v2.1.1 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/tls/2.1.1/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/tls v3.1.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/tls/3.1.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/vsphere v1.13.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/vsphere/1.13.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/vsphere v1.18.3 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/vsphere/1.18.3/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/vsphere v1.25.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/hashicorp/vsphere/1.25.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/ibm-cloud/ibm v0.31.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/ibm-cloud/ibm/0.31.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/ibm-cloud/ibm v1.14.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/ibm-cloud/ibm/1.14.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/nutanix/nutanix v1.2.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/nutanix/nutanix/1.2.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/tencentcloudstack/tencentcloud v1.39.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/tencentcloudstack/tencentcloud/1.39.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/terraform-provider-openstack/openstack v1.17.1 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/terraform-provider-openstack/openstack/1.17.1/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/terraform-provider-openstack/openstack v1.34.1 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/terraform-provider-openstack/openstack/1.34.1/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/vmware/nsxt v3.1.0 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/vmware/nsxt/3.1.0/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/vmware/nsxt v3.1.1 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/vmware/nsxt/3.1.1/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/vmware/vra7 v1.0.2 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/vmware/vra7/1.0.2/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/vmware/vra7 v2.0.1 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/vmware/vra7/2.0.1/linux_amd64 2021/05/25 20:08:35 
[TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/vmware/vra7 v3.0.2 for linux_amd64 at /home/terraform/.terraform.d/plugins/registry.terraform.io/vmware/vra7/3.0.2/linux_amd64 2021/05/25 20:08:35 
[DEBUG] ignoring non-existing provider search directory /stacks/stackjobs/temp_a29c2780-bd78-11eb-b4e9-b591e3e27225/.local/share/terraform/plugins 2021/05/25 20:08:35 
[DEBUG] ignoring non-existing provider search directory /usr/local/share/terraform/plugins 2021/05/25 20:08:35 
[DEBUG] ignoring non-existing provider search directory /usr/share/terraform/plugins 2021/05/25 20:08:35 
[INFO] Checkpoint disabled. Not running. 2021/05/25 20:08:35 
[INFO] CLI command args: []string{"init", "-backend=false", "-input=false", "-get=true"} 2021/05/25 20:08:35 
[TRACE] Meta.Backend: backend has not previously been initialized in this working directory 2021/05/25 20:08:35 
[TRACE] backend/local: state manager for workspace "default" will: - read initial snapshot from terraform.tfstate - write new snapshots to terraform.tfstate - create any backup at terraform.tfstate.backup 2021/05/25 20:08:35 
[TRACE] statemgr.Filesystem: reading initial snapshot from terraform.tfstate 2021/05/25 20:08:35 
[TRACE] statemgr.Filesystem: snapshot file has nil snapshot, but that's okay 2021/05/25 20:08:35 [TRACE] statemgr.Filesystem: read nil snapshot 
Initializing provider plugins...
- Finding hashicorp/google versions matching "3.69.0"...
Error: 
Failed to query available provider packages
Could not retrieve the list of available versions for provider hashicorp/google: no available releases match the given constraints 3.69.0
```
### Expected Behavior

Should download google provider 3.69.0  even if the older version of the provider is found locally. This is how engine used to behave in 0.11.x and 0.12.x.

### Actual Behavior

Fails to download the google provider 3.69.0 and throws the following error

```
Error: 
Failed to query available provider packages
Could not retrieve the list of available versions for provider hashicorp/google: no available releases match the given constraints 3.69.0
``` 
### Steps to Reproduce

1. Have a older version of google provider locally $HOME/.terraform.d/plugins/registry.terraform.io/hashicorp/google
2. terraform init on the configuration file.

### Additional Context

It looks like the issue is with how the automatic _exclude_ is computed (in the implicit implied _direct_ block) when provider_installation block is not in CLI configuration. Probably the exclude list computation must be based on the provider version rather than just the host/namespace/provider_type. 

In 0.12.x and 0.11.x if provider version 1.x.x is found in plugins directory locally, engine was able to download newer or older version based on the version setting in provider stanza.

If I have a explicit _provider_installation_ CLI configuration setting as follows then the new version is downloaded. But when older version is used in terraform configuration file, the locally available older version is ignored and is downloaded again which is waste of network traffic.  

```
provider_installation {
  direct {
  }
}
```

**Comment by mildwonkey** [comment]
Hi @bhadrim,
Terraform v0.13 introduced a [provider dependency lock file](https://www.terraform.io/docs/language/dependency-lock.html). If that file is present and the locked provider versions are present, terraform will not download newer providers. 

To install newer versions of providers, you can do the following:
* pin providers to the exact version you want, and update that when you want newer versions (recommended)
* run `terraform init -upgrade` to get the latest versions 
* delete the lock file and run `init` again 

Since this is expected behavior, I am going to close this issue. If you have any more questions about provider installation, I'd suggest asking in [the community forums](https://discuss.hashicorp.com/c/terraform-core/27). Thanks!

**Comment by bhadrim** [comment]
@mildwonkey Thank you for your reply. I believe your solution may not address my issue. I never ran init on the template with older provider version. I believe the lock file gets created only when the template is initialized. 

This is what I did: I pre-popuated the older versions of the providers in my plugins/registry.terraform.io/hashicorp/google and I created a new template with provider version 3.69.0 and when I run init I get the error mentioned in the original post.

I will post my question in community. Thank you.

**Comment by jbardin** [comment]
Thanks @bhadrim, I confirmed that the existence of local mirrors are superseding searching for newer versions in the registry.

**Comment by apparentlymart** [comment]
Hi @bhadrim,

I think we had a discussion about this in [a community forum topic](https://discuss.hashicorp.com/t/terraform-13-14-fails-to-download-newer-provider-version-if-older-version-exists-locally/25168/2) earlier.

As I noted there, the behavior you are seeing is as intended, and this was an intentional change made in Terraform v0.13 based on feedback that various people found it surprising that Terraform would try to download a remote version of a provider when the user had (as far as they were concerned) told Terraform to use a local version instead.

I shared some ideas in that community forum post for different ways to get the result you intended. Since the change we're discussing here is now quite old we would not be able to revert it (that would be a further breaking change, which we are now committed not to make) and so I'm going to close this issue. However, I'm happy to continue talking with you in the community forum topic if you have any follow-up questions about the suggestions I made there. Thanks!
